### PR TITLE
AutoActivation API Surface

### PR DESCRIPTION
**Pull request by geeknoid**
Please check out the APIs marked as [Experimental].

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/extensions/pull/4374)

### CODE REVIEW

File: API.Microsoft.Extensions.DependencyInjection.AutoActivation/WithDocs/Microsoft.Extensions.DependencyInjection/AutoActivationExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+namespace Microsoft.Extensions.DependencyInjection;
+
+/// <summary>
+/// Extension methods for automatically activating singletons after application starts.
+/// </summary>
+public static class AutoActivationExtensions
+{
+    /// <summary>
+    /// Enforces singleton activation at startup time rather then at runtime.
```
Comment (xakep139) [inline API.Microsoft.Extensions.DependencyInjection.AutoActivation/WithDocs/Microsoft.Extensions.DependencyInjection/AutoActivationExtensions.cs:14]: Can we add a remark that this method and other overloads have a prerequisite that a singleton is already registered in DI?

File: API.Microsoft.Extensions.DependencyInjection.AutoActivation/WithDocs/Microsoft.Extensions.DependencyInjection/AutoActivationExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    /// <param name="serviceKey">An object used to uniquely identity the specific service.</param>
+    /// <returns>The value of <paramref name="services" />.</returns>
+    [Experimental("EXTEXP0012", UrlFormat = "https://aka.ms/dotnet-extensions-warnings/{0}")]
+    public static IServiceCollection ActivateKeyedSingleton(this IServiceCollection services, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type serviceType, object? serviceKey);
+
+    /// <summary>
+    /// Adds an auto-activated keyed singleton service.
+    /// </summary>
+    /// <param name="services">The service collection to add the service to.</param>
+    /// <param name="serviceKey">An object used to uniquely identity the specific service.</param>
```
Comment (xakep139) [inline API.Microsoft.Extensions.DependencyInjection.AutoActivation/WithDocs/Microsoft.Extensions.DependencyInjection/AutoActivationExtensions.cs:164]: to identity -> to identify?

File: API.Microsoft.Extensions.DependencyInjection.AutoActivation/WithDocs/Microsoft.Extensions.DependencyInjection/AutoActivationExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    public static void TryAddActivatedKeyedSingleton(this IServiceCollection services, Type serviceType, object? serviceKey, Type implementationType);
+
+    /// <summary>
+    /// Tries to add an auto-activated keyed singleton service.
+    /// </summary>
+    /// <param name="services">The service collection to add the service to.</param>
+    /// <param name="serviceType">The type of the service to register.</param>
+    /// <param name="serviceKey">An object used to uniquely identity the specific service.</param>
+    /// <param name="implementationFactory">The factory that creates the service.</param>
+    [Experimental("EXTEXP0012", UrlFormat = "https://aka.ms/dotnet-extensions-warnings/{0}")]
```
Comment (danmoseley) [inline API.Microsoft.Extensions.DependencyInjection.AutoActivation/WithDocs/Microsoft.Extensions.DependencyInjection/AutoActivationExtensions.cs:262]: @jeffhandley is there a central doc where we have to claim/register ID's for [Experimental ] attributes? I thought we had one for BCLnnnn but I can't find that right now.

Comment (jeffhandley) [reply API.Microsoft.Extensions.DependencyInjection.AutoActivation/WithDocs/Microsoft.Extensions.DependencyInjection/AutoActivationExtensions.cs:262]: @danmoseley I believe you're thinking of this doc that captures the list of diagnostics (including obsoletions). The page also includes instructions for how we keep the constants organized in the dotnet/runtime codebase.

https://github.com/dotnet/runtime/blob/main/docs/project/list-of-diagnostics.md

We don't have a registration across repos, but dotnet/extensions could introduce the same approach that document promotes. You can also use CODEOWNERS to tag specific team members when new entries are added if desired.

Comment (RussKie) [reply API.Microsoft.Extensions.DependencyInjection.AutoActivation/WithDocs/Microsoft.Extensions.DependencyInjection/AutoActivationExtensions.cs:262]: We have a simple version of that doc too at https://github.com/dotnet/extensions/blob/main/docs/list-of-diagnostics.md.

Comment (RussKie) [reply API.Microsoft.Extensions.DependencyInjection.AutoActivation/WithDocs/Microsoft.Extensions.DependencyInjection/AutoActivationExtensions.cs:262]: A while ago I tried to see if there was one central registry for obsoletions and other diagnostics, and I was told that every repo/product maintains their own. 
To list a few:
* https://github.com/dotnet/runtime/blob/main/docs/project/list-of-diagnostics.md?rgh-link-date=2023-09-12T00%3A25%3A27Z
* https://github.com/dotnet/aspnetcore/blob/main/docs/list-of-diagnostics.md
* https://github.com/dotnet/winforms/blob/main/docs/list-of-diagnostics.md

**Pull request closed without merging**